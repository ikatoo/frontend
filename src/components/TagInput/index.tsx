import { InputHTMLAttributes, useState } from 'react'

import { Close } from '@styled-icons/material-outlined'
import Styles from './styles'

export type TagInputProps = {
  onTagsChange: (value: string[]) => void
  label?: string
  labelColor?: 'black' | 'white'
  initialTags?: string[]
  disabled?: boolean
  error?: string
  name: string
} & InputHTMLAttributes<HTMLInputElement>

const TagInput = ({
  label,
  labelColor = 'black',
  name,
  initialTags = [],
  error,
  disabled = false,
  onTagsChange,
  ...props
}: TagInputProps) => {
  const [inputValue, setInputValue] = useState('')

  const addTag = (tag: string) => {
    if (tag.length && !initialTags.includes(tag)) {
      onTagsChange([...initialTags, tag])
    }
    setInputValue('')
  }

  const removeTag = (tag: string) => {
    const newTags = initialTags.filter((_tag) => _tag !== tag)
    onTagsChange(newTags)
  }

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    if (newValue.endsWith(',')) {
      addTag(newValue.split(',')[0])
    } else {
      setInputValue(newValue)
    }
  }

  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key.toLowerCase() === 'backspace' && !inputValue.length) {
      const newValue = initialTags.slice(0, initialTags.length - 1)
      onTagsChange(newValue)
    }
  }

  return (
    <Styles.Wrapper disabled={disabled} error={!!error}>
      {!!label && (
        <Styles.Label labelColor={labelColor} htmlFor={name}>
          {label}
        </Styles.Label>
      )}
      <Styles.InputWrapper>
        <Styles.TagsWrapper>
          {initialTags.length > 0 &&
            initialTags.map((tag) => (
              <Styles.Tag data-testid="tag-test-id" key={tag}>
                {tag}
                <Styles.CloseWrapper onClick={() => removeTag(tag)}>
                  <Close size={15} />
                </Styles.CloseWrapper>
              </Styles.Tag>
            ))}
        </Styles.TagsWrapper>
        <Styles.Input
          type="text"
          onChange={onChange}
          onKeyDown={onKeyDown}
          value={inputValue}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </Styles.InputWrapper>
      {!!error && <Styles.Error>{error}</Styles.Error>}
    </Styles.Wrapper>
  )
}

export default TagInput
