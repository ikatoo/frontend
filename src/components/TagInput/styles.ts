import styled, { css } from 'styled-components'

import { TagInputProps } from '.'
import theme from '../../theme'
import tw from 'tailwind-styled-components/dist/tailwind'

type LabelProps = Pick<TagInputProps, 'labelColor'>

type WrapperProps = Pick<TagInputProps, 'disabled'> & { error?: boolean }

const InputWrapper = styled.div`
  ${() => css`
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

const Input = styled.input`
  color: ${theme.colors.black};
  font-family: ${theme.font.family};
  font-size: ${theme.font.sizes.large};
  padding: ${theme.spacings.xsmall} 0;
  padding-left: ${theme.spacings.xsmall};
  background: transparent;
  border: 0;
  outline: none;
  max-width: 5rem;
  margin: 0.2rem;
  border-style: dashed;
  border-width: thin;
  border-radius: ${theme.border.radius};
  border-color: transparent;

  &:-webkit-autofill {
    -webkit-box-shadow: 0 0 0 ${theme.spacings.small} ${theme.colors.lightGray}
      inset;
    filter: none;
    &::first-line {
      font-family: ${theme.font.family};
      font-size: ${theme.font.sizes.large};
    }
  }

  &:focus {
    border-color: ${theme.colors.gray};
  }
`

const Label = styled.label<LabelProps>`
  ${({ labelColor }) => css`
    font-size: ${theme.font.sizes.xxlarge};
    line-height: ${theme.spacings.xxxlarge};
    color: ${theme.colors[labelColor ?? 'black']};
    cursor: pointer;
  `}
`

const Error = styled.p`
  ${() => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

const wrapperModifiers = {
  error: () => css`
    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }

    ${Label} {
      color: ${theme.colors.red};
    }
  `,
  disabled: () => css`
    ${Label},
    ${Input} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `
}

const TagsWrapper = tw.div`
  flex
  flex-wrap
  gap-1
`

const Tag = styled.div`
  display: flex;
  padding: 0.3rem;
  background-color: ${theme.colors.gray};
  color: ${theme.colors.lightGray};
  border-radius: ${theme.border.radius};

  svg {
    margin: auto;
    padding-left: 0.2rem;
  }
`

const Wrapper = styled.div<WrapperProps>`
  ${({ error, disabled }) => css`
    ${error && wrapperModifiers.error()}
    ${disabled && wrapperModifiers.disabled()}
  `}
`

const CloseWrapper = tw.span`cursor-pointer`

export default {
  Wrapper,
  Error,
  Label,
  Input,
  InputWrapper,
  TagsWrapper,
  Tag,
  CloseWrapper
}
