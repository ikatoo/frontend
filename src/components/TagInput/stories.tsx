import { ComponentMeta, ComponentStory } from '@storybook/react'

import TagInput from '.'

export default {
  title: 'Components/TagInput',
  component: TagInput,
  args: {
    label: 'Skills',
    name: 'skills',
    initialValue: '',
    placeholder: 'skill1',
    disabled: false
  }
} as ComponentMeta<typeof TagInput>

export const Default: ComponentStory<typeof TagInput> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TagInput {...args} />
  </div>
)

export const withError: ComponentStory<typeof TagInput> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TagInput {...args} />
  </div>
)
withError.args = {
  error: 'Ops...something is wrong'
}
