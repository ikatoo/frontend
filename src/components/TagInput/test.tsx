import userEvent from '@testing-library/user-event'

import TagInput from '.'
import { act, render, screen, waitFor } from '@testing-library/react'
import { vi } from 'vitest'

describe('<TagInput />', () => {
  it('Renders with Label', () => {
    render(<TagInput label="Label" name="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    render(<TagInput name="username" />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    render(<TagInput name="username" placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onInputChange = vi.fn()
    render(
      <TagInput
        onInputChange={onInputChange}
        label="TagInput"
        name="TagInput"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'

    act(() => {
      userEvent.type(input, text)
    })

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onInputChange).toHaveBeenCalledWith(text)
  })

  it('Does not changes its value when disabled', async () => {
    const onInputChange = vi.fn()
    render(
      <TagInput
        onInputChange={onInputChange}
        label="TagInput"
        name="TagInput"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'

    act(() => {
      userEvent.type(input, text)
    })

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInputChange).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = render(
      <TagInput name="username" label="TagInput" error="Error message" />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    render(<TagInput label="TagInput" name="TagInput" />)

    const input = screen.getByLabelText('TagInput')
    expect(document.body).toHaveFocus()

    act(() => {
      userEvent.tab()
    })

    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(<TagInput label="TagInput" name="TagInput" disabled />)

    const input = screen.getByLabelText('TagInput')
    expect(document.body).toHaveFocus()

    act(() => {
      userEvent.tab()
    })

    expect(input).not.toHaveFocus()
  })

  it('When typing comma circul the text with gray background up to previous comma', () => {
    render(<TagInput label="TagInput" name="TagInput" />)
    const input = screen.getByLabelText('TagInput')

    act(() => {
      input.focus()
    })
    expect(input).toHaveFocus()

    userEvent.type(input, 'skill typed,')
  })
})
