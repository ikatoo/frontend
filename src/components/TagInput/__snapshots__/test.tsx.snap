// Vitest Snapshot v1

exports[`<TagInput /> > Renders with error 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #e5e7eb;
  border-radius: 0.2rem;
  padding: 0 0.4rem;
  border: 0.2rem solid;
  border-color: #e5e7eb;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #1e293b;
}

.c5 {
  color: #00021f;
  font-family: 'Oswald','Helvetica','Arial','sans-serif';
  font-size: 0.9rem;
  padding: 0.4rem 0;
  padding-left: 0.4rem;
  background: transparent;
  border: 0;
  outline: none;
  max-width: 5rem;
  margin: 0.2rem;
  border-style: dashed;
  border-width: thin;
  border-radius: 0.4rem;
  border-color: transparent;
}

.c5:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 0.6rem #e5e7eb inset;
  -webkit-filter: none;
  filter: none;
}

.c5:-webkit-autofill::first-line {
  font-family: 'Oswald','Helvetica','Arial','sans-serif';
  font-size: 0.9rem;
}

.c5:focus {
  border-color: #8F8F8F;
}

.c2 {
  font-size: 1.4rem;
  line-height: 2rem;
  color: #00021f;
  cursor: pointer;
}

.c6 {
  color: #ff2a2a;
  font-size: 0.6rem;
}

.c0 .c3 {
  border-color: #ff2a2a;
}

.c0 .c1 {
  color: #ff2a2a;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="username"
  >
    TagInput
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="flex flex-wrap gap-1"
    />
    <input
      class="c5"
      id="username"
      name="username"
      type="text"
      value=""
    />
  </div>
  <p
    class="c6"
  >
    Error message
  </p>
</div>
`
